# 03 - Dodanie Dodatkowych Features

## Cel: 
Uzupełnienie danych o cechy potrzebne do dalszej analizy danych i zależności między poziomem wody

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
os.chdir('..')

from src.data.preprocessing import build_main_df
final = build_main_df()
print("Kształt danych:", final.shape)
final.head()

Kształt danych: (1826, 34)


,Opad_suma,Temp_średnia,Temp_min,Temp_max,Wilgotność_średnia,Ciśnienie_średnia,Ciśnienie_min,Ciśnienie_max,month,sezon,...,Wiatr_cos_proxy,Opad_lag_1d,Opad_lag_2d,Opad_lag_3d,Temp_średnia_lag_1d,Opad_24h,Opad_72h,Opad_7d,Poziom_wody_średnia,Poziom_wody_max
Data,,,,,,,,,,,,,,,,,,,,,
2021-01-01,0.00,-1.170833,-2.2,-0.2,94.341667,1005.466667,1005.0,1006.6,1,zima,...,0.0,0.00,0.00,0.0,0.000000,0.00,2.08,28.35,-0.052917,0.00
2021-01-02,0.00,0.616667,-0.3,2.3,87.733333,1011.208333,1007.0,1015.7,1,zima,...,1.0,0.00,0.00,0.0,-1.170833,0.00,2.08,28.35,-0.035833,0.00
2021-01-03,2.08,0.558333,-0.8,1.6,91.033333,1016.716667,1015.9,1017.4,1,zima,...,1.0,0.00,0.00,0.0,0.616667,2.08,2.08,28.35,0.013333,0.05
2021-01-04,3.61,1.245833,0.6,1.9,90.575000,1016.441667,1015.9,1016.9,1,zima,...,-1.0,2.08,0.00,0.0,0.558333,3.61,5.69,28.35,0.015417,0.06
2021-01-05,18.58,2.170833,0.9,3.6,91.354167,1013.233333,1009.3,1016.3,1,zima,...,-1.0,3.61,2.08,0.0,1.245833,18.58,24.27,28.35,0.018750,0.10


## 1. Cechy sezonowe

In [2]:
print("Kolumny sezonowe:")
print(final[['month', 'sezon', 'doy', 'sin_doy', 'cos_doy']].head(10))

Kolumny sezonowe:
            month sezon  doy   sin_doy   cos_doy
Data                                            
2021-01-01      1  zima    1  0.017213  0.999852
2021-01-02      1  zima    2  0.034422  0.999407
2021-01-03      1  zima    3  0.051620  0.998667
2021-01-04      1  zima    4  0.068802  0.997630
2021-01-05      1  zima    5  0.085965  0.996298
2021-01-06      1  zima    6  0.103102  0.994671
2021-01-07      1  zima    7  0.120208  0.992749
2021-01-08      1  zima    8  0.137279  0.990532
2021-01-09      1  zima    9  0.154309  0.988023
2021-01-10      1  zima   10  0.171293  0.985220


#### Opis stworzonych cech sezonowych

- **month** — numer miesiąca (1-12), do mapowanie pory roku
- **sezon** — pora roku (zima/wiosna/lato/jesien), pozwala analizować 
  zależności osobno dla każdej pory roku
- **doy** — numer dnia roku (1-365)
- **sin_doy / cos_doy** — cykliczna reprezentacja dnia roku. 
  Zwykły numer miesiąca sugeruje że styczeń(1) i grudzień(12) są daleko 
  od siebie, a tak naprawde obok siebie. Sin/cos pokazuje tę cykliczność.

## 2. Cechy ciśnienia

In [3]:
print("Kolumny ciśnienia:")
print(final[['Ciśnienie_średnia', 'Ciśnienie_min', 'Ciśnienie_max','Ciśnienie_ampl','Ciśnienie_delta_1d','Ciśnienie_delta_2d','Ciśnienie_delta_3d','Ciśnienie_trend_3d', 'Ciśnienie_trend_7d']].head(10))

Kolumny ciśnienia:
            Ciśnienie_średnia  Ciśnienie_min  Ciśnienie_max  Ciśnienie_ampl  \
Data                                                                          
2021-01-01        1005.466667         1005.0         1006.6             1.6   
2021-01-02        1011.208333         1007.0         1015.7             8.7   
2021-01-03        1016.716667         1015.9         1017.4             1.5   
2021-01-04        1016.441667         1015.9         1016.9             1.0   
2021-01-05        1013.233333         1009.3         1016.3             7.0   
2021-01-06        1007.900000         1006.5         1009.2             2.7   
2021-01-07        1004.550000         1003.6         1006.2             2.6   
2021-01-08        1006.983333         1005.0         1009.7             4.7   
2021-01-09        1012.500000         1009.9         1015.9             6.0   
2021-01-10        1016.608333         1015.1         1017.8             2.7   

            Ciśnienie_delta_1d  

#### Opis stworzonych cech ciśnienia

- **Ciśnienie_średnia / min / max** — dobowe statystyki ciśnienia
- **Ciśnienie_ampl** — amplituda ciśnienia w ciągu dnia (max - min)
- **Ciśnienie_delta_1d / 2d / 3d** — zmiana ciśnienia względem 1/2/3 dni wcześniej
- **Ciśnienie_trend_3d / 7d** — średnia ciśnienia z 3 i 7 dni

## 3. Proxy wiatru

In [4]:
print("Proxy wiatru:")
print(final[['Wiatr_kierunek_proxy', 'Wiatr_sektor_proxy', 
             'Wiatr_sin_proxy', 'Wiatr_cos_proxy']].head(10))
print("\nRozkład sektorów:")
print(final['Wiatr_sektor_proxy'].value_counts())

Proxy wiatru:
            Wiatr_kierunek_proxy Wiatr_sektor_proxy  Wiatr_sin_proxy  \
Data                                                                   
2021-01-01                   0.0          stabilnie     0.000000e+00   
2021-01-02                   1.0   wzrost_cisnienia     0.000000e+00   
2021-01-03                   1.0   wzrost_cisnienia     0.000000e+00   
2021-01-04                  -1.0   spadek_cisnienia     1.224647e-16   
2021-01-05                  -1.0   spadek_cisnienia     1.224647e-16   
2021-01-06                  -1.0   spadek_cisnienia     1.224647e-16   
2021-01-07                  -1.0   spadek_cisnienia     1.224647e-16   
2021-01-08                   1.0   wzrost_cisnienia     0.000000e+00   
2021-01-09                   1.0   wzrost_cisnienia     0.000000e+00   
2021-01-10                   1.0   wzrost_cisnienia     0.000000e+00   

            Wiatr_cos_proxy  
Data                         
2021-01-01              0.0  
2021-01-02              1.0  
2

#### Opis stworzonych kolumn z proxy

**Nie sa to prawdziwe dane o wietrze, zostały one wygenerowane na podstawie amplitudy cisnienia**
- **Wiatr_siła_proxy** — bezwzględna zmiana ciśnienia
- **Wiatr_kierunek_proxy** — znak zmiany (-1/0/1)
- **Wiatr_sektor_proxy** — kategoria: spadek/stabilnie/wzrost
- **Wiatr_kąt_proxy / sin / cos** — reprezentacja kierunku w postaci liczby

## 4.  Lagi

In [5]:
print("Lagi opadów:")
print(final[['Opad_suma', 'Opad_lag_1d', 'Opad_lag_2d', 'Opad_lag_3d','Temp_średnia_lag_1d']].head(10))

Lagi opadów:
            Opad_suma  Opad_lag_1d  Opad_lag_2d  Opad_lag_3d  \
Data                                                           
2021-01-01       0.00         0.00         0.00         0.00   
2021-01-02       0.00         0.00         0.00         0.00   
2021-01-03       2.08         0.00         0.00         0.00   
2021-01-04       3.61         2.08         0.00         0.00   
2021-01-05      18.58         3.61         2.08         0.00   
2021-01-06       3.31        18.58         3.61         2.08   
2021-01-07       0.77         3.31        18.58         3.61   
2021-01-08       0.00         0.77         3.31        18.58   
2021-01-09       1.91         0.00         0.77         3.31   
2021-01-10       0.00         1.91         0.00         0.77   

            Temp_średnia_lag_1d  
Data                             
2021-01-01             0.000000  
2021-01-02            -1.170833  
2021-01-03             0.616667  
2021-01-04             0.558333  
2021-01-05    

#### Opis kolumn z lagami

- **Opad_lag_1d / 2d / 3d** — opad z poprzednich dni
- **Temp_średnia_lag_1d** — temperatura z dnia poprzedniego

## 5.  Kolumny z skumulowanym opadem

In [6]:
print("Opady skumulowane:")
print(final[['Opad_suma', 'Opad_24h', 'Opad_72h', 'Opad_7d']].head(10))

Opady skumulowane:
            Opad_suma  Opad_24h  Opad_72h  Opad_7d
Data                                              
2021-01-01       0.00      0.00      2.08    28.35
2021-01-02       0.00      0.00      2.08    28.35
2021-01-03       2.08      2.08      2.08    28.35
2021-01-04       3.61      3.61      5.69    28.35
2021-01-05      18.58     18.58     24.27    28.35
2021-01-06       3.31      3.31     25.50    28.35
2021-01-07       0.77      0.77     22.66    28.35
2021-01-08       0.00      0.00      4.08    28.35
2021-01-09       1.91      1.91      2.68    30.26
2021-01-10       0.00      0.00      1.91    28.18


#### Opis kolumn z skumulowanymi opadami

### Opady skumulowane
- **Opad_24h** — opad bieżącego dnia
- **Opad_72h** — suma z 3 dni (proxy nasycenia zlewni)
- **Opad_7d** — suma z 7 dni

## 5.  Kolumny z poziomem wody

In [7]:
print("Dane poziomu wody:")
print(final[['Poziom_wody_średnia', 'Poziom_wody_max']].head(10))
print("\nStatystyki:")
print(final[['Poziom_wody_średnia', 'Poziom_wody_max']].describe())

Dane poziomu wody:
            Poziom_wody_średnia  Poziom_wody_max
Data                                            
2021-01-01            -0.052917             0.00
2021-01-02            -0.035833             0.00
2021-01-03             0.013333             0.05
2021-01-04             0.015417             0.06
2021-01-05             0.018750             0.10
2021-01-06            -0.030000             0.04
2021-01-07            -0.072917            -0.05
2021-01-08            -0.081667            -0.04
2021-01-09            -0.058750            -0.02
2021-01-10            -0.189583            -0.10

Statystyki:
       Poziom_wody_średnia  Poziom_wody_max
count          1826.000000      1826.000000
mean              0.084264         0.171750
std               0.179736         0.228293
min              -0.481667        -0.390000
25%              -0.033229         0.040000
50%               0.078542         0.160000
75%               0.188229         0.280000
max               1.325333  

Dane poziomu wody pochodzą z czujnika na rzece Strzyża 
i zostały połączone z danymi meteorologicznymi po dacie.

- **Poziom_wody_średnia** - średni dobowy poziom wody [m]
- **Poziom_wody_max** - maksymalny dobowy poziom wody [m]

## Podsumowanie

In [8]:
print("Kształt finalnego datasetu:", final.shape)
print(f"\nOkres: {final.index.min().date()} → {final.index.max().date()}")
print(f"Liczba dni: {len(final)}")
print(f"\nLiczba cech: {len(final.columns)}")
print("\nWszystkie kolumny:")
print(final.columns.tolist())

Kształt finalnego datasetu: (1826, 34)

Okres: 2021-01-01 → 2025-12-31
Liczba dni: 1826

Liczba cech: 34

Wszystkie kolumny:
['Opad_suma', 'Temp_średnia', 'Temp_min', 'Temp_max', 'Wilgotność_średnia', 'Ciśnienie_średnia', 'Ciśnienie_min', 'Ciśnienie_max', 'month', 'sezon', 'doy', 'sin_doy', 'cos_doy', 'Ciśnienie_ampl', 'Ciśnienie_delta_1d', 'Ciśnienie_delta_2d', 'Ciśnienie_delta_3d', 'Ciśnienie_trend_3d', 'Ciśnienie_trend_7d', 'Wiatr_siła_proxy', 'Wiatr_kierunek_proxy', 'Wiatr_sektor_proxy', 'Wiatr_kąt_proxy', 'Wiatr_sin_proxy', 'Wiatr_cos_proxy', 'Opad_lag_1d', 'Opad_lag_2d', 'Opad_lag_3d', 'Temp_średnia_lag_1d', 'Opad_24h', 'Opad_72h', 'Opad_7d', 'Poziom_wody_średnia', 'Poziom_wody_max']


Dataset zawiera 1826 dni pomiarowych (2021-2025) z 34 kolumnami:
- 8 surowych zmiennych meteorologicznych
- 5 cech sezonowych
- 10 cech ciśnienia i proxy wiatru
- 4 lagi czasowe
- 3 opady skumulowane
- 2 zmienne poziomu wody (Poziom_wody_średnia, Poziom_wody_max)